include/sparrow/fixed_size_layout.hpp

```diff
@@ -121,7 +121,7 @@ namespace sparrow
         // using iterator = layout_iterator<self_type, false>;
         // using const_iterator = layout_iterator<self_type, true>;
 
-        explicit fixed_size_layout(array_data p);
+        explicit fixed_size_layout(array_data& data);
```

Comment:
In general, can layouts be moved or copied? My assumption is yes but if so we need to add ways for these operations to be safe, in particular in the context of moving `array_data` objects around (if that would happen). Because now that it's a pointer being stored, moved-from layouts might lead to UB?

Reply:
Ah good catch, the overall structure would be something like:
```cpp
tempate < .... >
class typed_array
{

private:
    array_data m_data;
    layout_type m_layout;
];
```
With m_layout referencing `m_data`, and everything being movable / copiable ...

Reply:
One solution would be to move `m_data` into a `std::unique_ptr` to guarantee in-memory stability.

Reply:
A third option is to delete copy and move semantics of layouts and reinstantiate them each time we copy / move a `typed_array`. This is actually my favorite option.

Reply:
So that leads to questions like:
- if we put a `unique_ptr` in `typed_array` , is the cost of allocation a problem? (assuming both members would be contained in the allocated object, like a pimpl)
- is it ok to implement `typed_array`'s copy operations manually? what happens to `m_layout` if it's not copyable but we need to copy the `typed_array`?

I feel like keeping layouts copy and move enabled might be the simplest solution, not sure...

Reply:
The copy / move constructor of `typed_array` would be something like:
```cpp
typed_array(const typed_array& rhs)
    : m_data(rhs.m_data)
    , m_layout(m_data)
{
}

typed_array(typed_array&& rhs)
    : m_data(std::move(rhs.m_data))
    , m_layout(m_data)
{
}
```

The issue is for assignment operators (copy and move). We probably need an additional "reset_array_data" method to be able to implement these operators.

The issue with keeping the copy and move semantic is that it will be sliently wrong the layout ogf a copied array will still refer to the array_data of the initial array.

> if we put a unique_ptr in typed_array , is the cost of allocation a problem? (assuming both members would be contained in the allocated object, like a pimpl)

I don't really know; I would say that as much as we can avoid dynamic allocation, we should do it. Especially if the copy / move semantic of `typed_array` is "easy enough" to implement without the unique_ptr (even if I admit that "easy enough" is totally suggestive)

Reply:
> I would say that as much as we can avoid dynamic allocation, we should do it. 

I agree.

> The issue is for assignment operators (copy and move). We probably need an additional "reset_array_data" method to be able to implement these operators.

In those cases, only the `m_data` should be moved and copied, as `m_layout` always refers to the local `m_data`. No need to change `m_layout`... Or am I missing something? Does `m_layout` have an actual state? Depends on the layout type?

Reply:
The layout classes should be stateless. 

And yes, only the `array_data` member should be copied / moved. But then the layout member must be "built" again to refer the "new" array_data.